src/mlpack/tests/neat_test_tools.hpp

```diff
@@ -0,0 +1,155 @@
+/**
+ * @file neat_test_tools.hpp
+ * @author Rahul Ganesh Prabhu
+ *
+ * This file includes some useful classes for neat tests.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_TESTS_NEAT_TEST_TOOLS_HPP
+#define MLPACK_TESTS_NEAT_TEST_TOOLS_HPP
+
+#include <mlpack/core.hpp>
+#include <mlpack/methods/reinforcement_learning/environment/continuous_multiple_pole_cart.hpp>
+#include <mlpack/methods/reinforcement_learning/environment/multiple_pole_cart.hpp>
+#include <mlpack/methods/ann/activation_functions/hard_sigmoid_function.hpp>
+#include <mlpack/methods/ann/activation_functions/tanh_function.hpp>
+#include <mlpack/methods/neat/neat.hpp>
+
+using namespace mlpack;
+using namespace mlpack::neat;
+using namespace mlpack::ann;
+using namespace mlpack::rl;
+
+/**
+ * A class that defines the XOR task. 
+ */
+class XORTask
+{
+ public:
+  double Evaluate(Genome<HardSigmoidFunction>& genome)
+  {
+    arma::mat input = {{0, 0, 1, 1},
+                       {0, 1, 0, 1}};
+
+    double error = 0;
+    for (size_t i = 0; i < input.n_cols; i++)
+    {
+      arma::vec inputVec = input.col(i);
+      arma::vec output = genome.Evaluate(inputVec);
+      arma::vec answer = {(input(0, i) + input(1, i)) * (!input(0, i) +
+          !input(1, i))};
+      error += std::pow(answer[0] - output[0], 2);
+    }
+    return 4 - error;
+  }
+};
+
+/**
+ * A class that defines the non-Markovian double pole balancing task. The
+ * observation space is comprised of the cart position and the angles of the poles.
+ */
+class DPNVTask
+{
+ public:
+  DPNVTask(const MultiplePoleCart& env) : environment(env)
```

Comment:
I'm wondering if the implementation of the task is wrong, I can see there is a test for the env: https://github.com/mlpack/mlpack/blob/c62088ad40c4f9a3839e23a5dbbd5ab55054c7ee/src/mlpack/tests/rl_components_test.cpp#L171 but we haven't really tested if e.g. QLearning is able to solve the env, maybe you tested it?

Reply:
I think the learners we have right now aren't compatible with how the multiple pole cart class is structured. For instance, we use a matrix for the state data, and don't have a default constructor, for state and the environment. Maybe we should rewrite this class to allow for compatibility?

I've checked the equations again, and they are right according to the paper. I checked other implementations, and they are using RK4 along with these equations, which I don't know much about. I'll study about that and test that as well.

Reply:
We could work around the issue of not having a default constructor for the state by making the multiple pole cart into an only double pole cart method. That way the API will be intact without having an odd exception environment. Moreover, I guess I haven't seen a triple or quadruple pole balancing test before (I made it general since I thought I could). Unless there's a better workaround?

Reply:
> We could work around the issue of not having a default constructor for the state by making the multiple pole cart into an only double pole cart method. That way the API will be intact without having an odd exception environment. 

That sounds like a reasonable solution to me.

> Moreover, I guess I haven't seen a triple or quadruple pole balancing test before (I made it general since I thought I could). Unless there's a better workaround?

Yeah, I really liked that.

> I was also thinking, unless it would make running the tests too lengthy, we should add unit tests for all these environments, where we check if one of our learners could solve them, so that we don't run into this problem or the one we had with the pendulum environment. It would make sure the API remains consistent. Let me know what you think.

Had the same idea, we could provide some initial parameter that are almost good enough, that way we could reduce the runtime for the tests.

Reply:
I'll implement it locally for now, if NEAT works on it I'll open a PR to fix these issues.

Reply:
Sounds good, thanks!